File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.go
func canPartition(nums []int) bool {
  sum := 0
  mx := 0
  for _, v := range nums {
    sum += v
    if v > mx { mx = v }
  }  
  if sum % 2 == 1 || mx + mx > sum { return false }
  sum /= 2
  if mx == sum { return true }
  dp := make([]bool, sum+1)
  dp[0] = true
  for _, v := range nums {
    for j := sum; j >= v; j-- {
      if dp[j-v] {
        dp[j] = true
      }
    }
  }
  return dp[sum]
}

File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.rs
impl Solution {
  pub fn can_partition(nums: Vec<i32>) -> bool {
    let (max, sum) = (nums.iter().max().unwrap(), nums.iter().sum());
    if sum % 2 != 0 || max+max > sum { return false }
    let target = (sum / 2) as usize;
    let mut dp = vec![false; target + 1];
    dp[0usize] = true;
    for &n in nums.iter() {
      for i in (n as usize..=target).rev() {
        dp[i] = dp[i] || dp[i-(n as usize)]
      }
    }
    dp[target]
  }
}